src/java.desktop/share/classes/javax/swing/SwingWorker.java

```diff
@@ -753,6 +753,13 @@ public void run() {
         if (SwingUtilities.isEventDispatchThread()) {
             doDone.run();
         } else {
+            if (state != StateValue.DONE) {
+                do {
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {}
+                } while (state != StateValue.DONE);
+            }
```

Comment:
In the previous iteration, using `sleep` for waiting was *the concern*, you're still using `sleep`.

This is not going to work because it makes `cancel` wait until `DONE` state is reached, which is not what one would expect, especially taking into account that `cancel` is likely called from EDT to cancel the background operation and blocking EDT is not acceptable.

Reply:
I removed sleep from EDT case and not blocking EDT and guess sleep is now being done for non-EDT case...
If it is to be called from EDT then it should go to "if" block and not to "else" which is what I based my fix on..

Anyway, I appreciate your fix and will see to it..

Reply:
Yet having `sleep` is an indication of busy wait. You could've used `CountDownLatch` in conjunction with `setState` to avoid `sleep` at all.

> I removed sleep from EDT case and not blocking EDT and guess sleep is now being done for non-EDT case...

Ah, right. But then if `cancel` is called on EDT, `done` is invoked before `doInBackground` completes, isn't it?

If `cancel` is called from another thread, the main thread as in the test, it doesn't return until `doInBackground` completes. With your test case and fix, `cancel` blocks for 5 seconds.

Either way, the behaviour does not look right.

Reply:
It seems the isDone method modification is causing a JCK test to fail and there's no spec for it to challenge JCK test, although logically it seems right that STATE should be DONE but for now, I have modified the fix which satisfies regression test and JCK both..

Reply:
isDone() spec mention `"Completion may be due to normal termination, an exception, or cancellation -- in all of these cases, this method will return true"`
so for cancel state also, the state needs to be in DONE state, in which case we can reinstate the DONE state check in isDone()

Reply:
`isDone()` spec is inherited from `Future.isDone()` and it is true for how `FutureTask` is implemented.

https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L565-L570

Can't we change it?

Given what is said in the spec for `SwingWorker` before, I assume `isDone` should return `true` when the status of `SwingWorker` moves to `DONE` and it should occur when `doInBackground` returns as this is defined in the spec:

https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293